rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+  These types are handled by "generic" APIs whose behavior does not depend on
+  the value of each object (such as `tf.function`, `SavedModel`, and
+  `tf.while_loop`).
+
+* **_Tensor-like types_**, which specialize or extend tf.Tensor.
+  Values of these types have a `rank`, a `shape`, and usually a `dtype`.  In
+  addition to the "generic" APIs, these types can be handled by Tensor-specific
+  APIs (such as `tf.stack`, `tf.add`, and `tf.reduce_mean`).
+
+Examples of user-defined types that could defined or extended with this protocol
+include:
+
+**General data structures:**
+
+* `tfp.Distribution`: Encodes a statistical distribution.
+* `TensorDigraph`: Encodes the set of nodes and edges in a directed graph.
+* `DimensionAlignment`: Encodes a correspondence between two related dimensions
+  (e.g., between a `word` dimension and a `speaker` dimension).
+
+**Tensor-like types:**
+
+* `CSRSparseTensor`: A sparsely-encoded tensor that uses the Compressed Sparse
+  Row encoding.
+* `MaskedTensor`: Pairs a Tensor with a corresponding boolean mask, indicating
+  which values are valid, and automatically updates the mask as appropriate
+  when used with TensorFlow ops (such as `tf.add` or `tf.reduce_sum`).
+* `LabeledTensor`: Pairs a Tensor with a list of axis names, which can be used
+  for error detection and reporting, for specifying axes by name, and for
+  broadcasting.
+
+## User Benefit
+
+### Object-Oriented TensorFlow
+
+This proposal brings the benefits of Object-Oriented Programming to TensorFlow
+users, allowing them to define modular encapsulated data structures that
+interoperate with TensorFlow APIs.  This allows TensorFlow models to be defined
+at a higher level of abstraction.
+
+### Development outside the main TensorFlow repository
+
+Prior to this proposal, the only way to develop such data structures (e.g.
+`SparseTensor`) was to develop them inside the main TensorFlow code base.  This
+introduced significant barriers to rapid development, including slow release
+cycles, strong backwards compatibility constraints, and centralized API
+approval.  By allowing such data structures to be developed outside the main
+TensorFlow code base, we hope to make it much easier to experiment with new
+types and designs.  If general-purpose types are developed that become
+sufficiently mature, we may consider bringing them into the main TensorFlow code
+base.
+
+### TensorFlow APIs supported by user-defined types
+
+User-defined types that implement the interface proposed by this RFC will be
+supported by the following APIs:
+
+* **Keras**: User-defined types can be used as inputs and outputs for Keras
+  `Models` and `Layers`.
+* **tf.data.Dataset**: User-defined types can be included in `Datasets`, and
+  returned by dataset `Iterators`.
+* **Tensorflow hub**: User-defined types can be used as inputs and outputs for
+  `tf.hub` modules.
+* **SavedModel**: User-defined types can be used as inputs and outputs for
+  `SavedModels`.
```

Comment:
`SavedModel.signatures` intentionally flattens nested structures, because it needs to interoperate with non-Python interfaces (e.g. c++).  But if you just call the restored function on the `SavedModel` directly (rather than via the `signatures` dictionary), then you'll get proper handling of both nested values and extension types.  I.e., in Tensorflow/Tensorflow#40501, if you change `loaded_model.signatures['test'](inputs)` to `loaded_model.test(inputs)`, then it will handle both nested values and composite tensors (aka extension types) appropriately.

I do have a preliminary design for ExtensionTypes and TypeSpecs in c++ (there's a short blurb in the "Future Work" section), but I'm not including it in this RFC since there are still some details to work out.

Reply:
Understood. Thanks!